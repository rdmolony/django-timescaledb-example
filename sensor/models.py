import textwrap

from django.contrib.postgres.fields import ArrayField
from django.db import models


class FileType(models.Model):
    name = models.CharField(max_length=50)
    na_values = ArrayField(
        base_field=models.CharField(max_length=10),
        default=["NaN"],
        help_text=textwrap.dedent(
            r"""A list of strings to recognise as empty values. <br>
            
            Default: ["NaN"] <br>

            Note: "" is also included by default <br>

            Example: ["NAN", "-9999", "-9999.0"]
            """
        ),
    )
    delimiter = models.CharField(
        max_length=5,
        help_text=textwrap.dedent(
            r"""The character used to separate fields in the file. <br>
            
            Default: "," <br>
            
            Examples: "," or ";" or "\s+" for whitespace or "\t" for tabs
            """
        ),
        default=",",
    )
    datetime_fieldnames = ArrayField(
        base_field=models.CharField(max_length=50),
        default=["Tmstamp"],
        help_text=textwrap.dedent(
            r"""A list of datetime field names. <br>
            
            Examples: <br>
            
            1) Data has a single datetime field named "Tmstamp" which has values like
            '2021-06-29 00:00:00.000':  ["Tmstamp"] <br>

            2) Data has two datetime fields named "Date" and "Time" which have values
            like '01.01.1999' and '00:00' respectively: ["Date","Time"] <br>
            """
        ),
    )
    encoding = models.CharField(
        max_length=25,
        help_text=textwrap.dedent(
            r"""The encoding of the file. <br>

            Default: "utf-8" <br>

            Examples: utf-8 or latin-1 or cp1252
            """
        ),
        default="utf-8",
    )
    datetime_formats = ArrayField(
        base_field=models.CharField(max_length=25),
        help_text=textwrap.dedent(
            r"""The datetime format of `datetime_columns`. <br>

            See https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes
            for format codes

            Default: "%Y-%m-%d %H:%M:%S" <br>

            Examples: "%Y-%m-%d %H:%M:%S" for "2021-03-01 00:00:00"
            """
        ),
        default=[r"%Y-%m-%d %H:%M:%S"],
    )


class File(models.Model):
    file = models.FileField(upload_to="readings/", blank=False, null=False)
    uploaded_at = models.DateTimeField(auto_now_add=True)
    type = models.ForeignKey(FileType, on_delete=models.RESTRICT)


class Reading(models.Model):
    timestamp = models.DateTimeField(blank=False, null=False, primary_key=True)
    sensor_name = models.TextField(blank=False, null=False)
    reading = models.TextField(blank=False, null=False)

    class Meta:
        managed = False
