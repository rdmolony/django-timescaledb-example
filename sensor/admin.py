from django.contrib import admin
from import_export.admin import ImportExportModelAdmin
from import_export import resources

from .models import Reading


class ReadingResource(resources.ModelResource):

    class Meta:
        model = Reading

class ReadingAdmin(ImportExportModelAdmin):
    resource_classes = [ReadingResource]

admin.site.register(Reading, ReadingAdmin)
