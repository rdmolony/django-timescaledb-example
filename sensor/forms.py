from django.forms import ModelForm

from .models import File


class FileForm(ModelForm):
    class Meta:
        model = File
        fields = "__all__"
