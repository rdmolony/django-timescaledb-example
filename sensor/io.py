from ast import literal_eval
from collections import OrderedDict
from datetime import datetime
from datetime import timezone
import re
import typing

from django.core.exceptions import ValidationError

from .models import Reading


def unescape_backslash(s: str) -> str:
    """
    Unescape a backslash-escaped string like `\\t` to `\t`.

    Source: https://stackoverflow.com/questions/1885181/how-to-un-escape-a-backslash-escaped-string
    """
    return literal_eval(f'"{s}"')


def yield_readings(
    lines: typing.Iterator[str],
    delimiter: str,
    datetime_fieldnames: typing.Iterable[str],
    datetime_formats: typing.Iterable[str],
) -> typing.Iterator[typing.Tuple[typing.Any]]:
    """
    Yield wide-format lines as narrow-format readings

    https://en.wikipedia.org/wiki/Wide_and_narrow_data
    """
    split = lambda s: re.split(unescape_backslash(delimiter), s)
    standardised_lines = (split(line) for line in lines)

    fieldnames = None

    for line in standardised_lines:

        if set(datetime_fieldnames).issubset(set(line)):
            fieldnames = line
            break
    
    if fieldnames == None:
        raise ValidationError(f"No `datetime_fieldnames` {datetime_fieldnames} found!")

    for line in standardised_lines:
        fields = OrderedDict([(f, v) for f, v in zip(fieldnames, line)])
        readings = OrderedDict(
            [(f, v) for f, v in fields.items() if f not in datetime_fieldnames]
        )

        timestamp_strs = [fields[k] for k in datetime_fieldnames]
        timestamp_str = " ".join(
            str(item) for item in timestamp_strs if item is not None
        )

        for datetime_format in datetime_formats:
            try:
                timestamp = datetime.strptime(timestamp_str, datetime_format)
            except ValueError:
                pass
            else:
                for sensor, reading in readings.items():
                    yield {
                        "timestamp": timestamp, "sensor": sensor, "reading": reading,
                    }


def insert(
    rows: typing.Iterable[str],
    columns: typing.Iterable[str],
    table: str,
    primary_key_columns: typing.Iterable[str],
) -> None:

    n_rows_to_insert = 10_000
    _uuid = secrets.token_hex(8)
    temp_table = f"{table}__temp__{_uuid}"

    msg = (
        "`primary_key_columns` must be defined"
        + " in order to skip duplicate values on copying from a file!"
    )
    assert primary_key_columns, msg
    primary_keys = ", ".join(primary_key_columns)

    try:
        with transaction.atomic():
            with closing(connection.cursor()) as cursor:

                cursor.execute(
                    f"""
                    create temp table {temp_table} (
                        like {table}
                    );
                    """
                )

                cursor.copy_from(
                    file=StringIteratorIO(row for row in rows),
                    table=temp_table,
                    columns=columns,
                    sep=',',
                    null="NULL"
                )

                cursor.execute(f"select count(*) from {temp_table}")
                result = cursor.fetchone()
                table_size = result[0]
                assert table_size > 0, f"No readings found in file!"

                for offset in range(0, table_size, n_rows_to_insert):
                    percentage_complete = round(100 * offset / table_size, 2)
                    cursor.execute(
                        f"""
                        -- {percentage_complete}%
                        with chunk as (
                            select *
                            from {temp_table}
                            limit {n_rows_to_insert}
                            offset {offset}
                        )
                                insert into {table}
                            select distinct on ({primary_keys}) *
                                    from chunk;
                        """
                    )

                cursor.execute(
                    f"""
                    drop table {temp_table}
                    """
                )

    except Exception as e:
        raise ValidationError(f"File parsing failed  | Error: {e})") from e


def import_to_db(
    file_obj,
) -> None:

    target_table = "sensor_reading_generic"

    if not file_obj.type:
        message = (
            "Please define this file's type"
            + " before attempting to parse it"
            + " so the file's `encoding`, `delimiter`, `datetime_fieldnames`"
            + " etc are defined!"
        )
        raise ValidationError(message)

    try:

        # NOTE: ensure that the file datetime_format, encoding etc are valid!
        file_obj.clean()

        with file_obj.file.open(encoding=file_obj.type.encoding) as f:
            rows = yield_readings(
                lines=f,
                encoding=file_obj.type.encoding,
                delimiter=file_obj.type.delimiter,
                datetime_fieldnames=file_obj.type.datetime_fieldnames,
                datetime_formats=file_obj.type.datetime_formats,
                na_values=file_obj.type.na_values,
            )

        sql.insert(
            rows=rows,
            columns=columns,
            table=target_table,
            primary_key_columns=["station_id", "sensor_id", "timestamp"],
        )

    except Exception as e:
        file_obj.parsed = None
        file_obj.parse_error = str(e)
        file_obj.save()
        raise e

    else:
        file_obj.parsed = datetime.now(timezone.utc)
        file_obj.parse_error = None
        file_obj.save()
