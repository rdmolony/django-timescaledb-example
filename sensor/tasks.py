from celery import shared_task

from . import io


@shared_task
def import_to_db(file_obj):
    io.import_to_db(file_obj)
