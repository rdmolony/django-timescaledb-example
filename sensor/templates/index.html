<div style="text-align: center">
  <h1>django-timescaledb-example</h1>
  <div  style="margin-bottom: 10px">
    <button>
      <a href="{% url 'sensor:create-file-type' %}">Create File Type</a>
    </button>
  </div>
  <div style="margin-bottom: 10px">
    <button>
      <a href="{% url 'sensor:upload-file' %}">Upload File</a>
    </button>
  </div>
</div>