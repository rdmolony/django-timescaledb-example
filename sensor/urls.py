from django.shortcuts import redirect
from django.urls import include
from django.urls import path

from . import views


app_name = "sensor"


urlpatterns = [
    path('', lambda request: redirect('sensor:upload-file')),

    path('upload-file/', views.upload_file, name="upload-file"),
]
