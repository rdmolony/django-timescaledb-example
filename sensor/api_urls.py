from django.urls import include
from django.urls import path
from rest_framework.decorators import api_view
from rest_framework.response import Response
from rest_framework.reverse import reverse

from .api import viewsets


app_name = "sensor"


file_list = viewsets.FileViewSet.as_view({
    'get': 'list',
    'post': 'create'
})
file_detail = viewsets.FileViewSet.as_view({
    'get': 'retrieve',
    'put': 'update',
    'patch': 'partial_update',
    'delete': 'destroy'
})


@api_view(['GET'])
def api_root(request, format=None):
    return Response({
        'files': reverse('api:sensor:file-list', request=request, format=format),
        'echo': reverse('api:sensor:echo', request=request, format=format),
    })


urlpatterns = [
    path('', api_root, name="api-root"),
    path('file/', file_list, name='file-list'),
    path('file/<int:pk>/', file_detail, name='file-detail'),
]
