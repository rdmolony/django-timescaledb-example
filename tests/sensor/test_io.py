from datetime import datetime

import pytest

from sensor.io import unescape_backslash
from sensor.io import yield_readings


@pytest.mark.parametrize("string,expected", [("\\t", "\t")])
def test_unescape_backslash(string, expected):
    assert unescape_backslash(string) == expected


@pytest.mark.parametrize(
    "lines,delimiter,datetime_fieldnames,datetime_formats",
    [
        (
            iter(
                [
                    "Lat=0  Lon=0  Hub-Height=160  Timezone=00.0  Terrain-Height=0.0",
                    "Computed at 100 m resolution",
                    " ",
                    "YYYYMMDD HHMM   M(m/s) D(deg) SD(m/s)  DSD(deg)  Gust3s(m/s)    T(C)    PRE(hPa)       RiNumber  VertM(m/s)",
                    "20151222 0000  20.54   211.0    1.22       0.3        21.00     11.9      992.8            0.15    0.18",
                    "20151222 0010  21.02   212.2    2.55       0.6        21.35     11.8      992.7            0.29   -0.09",
                ]
            ),
            "\s+",
            ["YYYYMMDD", "HHMM"],
            [r"%Y%m%d %H%M"],
        ),
    ]
)
def test_yield_readings(
    lines,
    delimiter,
    datetime_fieldnames,
    datetime_formats,
    snapshot,
) -> None:
    output = [
        reading 
        for reading in yield_readings(
            lines=lines,
            delimiter=delimiter,
            datetime_fieldnames=datetime_fieldnames,
            datetime_formats=datetime_formats,
        )
    ]
    assert output == snapshot
