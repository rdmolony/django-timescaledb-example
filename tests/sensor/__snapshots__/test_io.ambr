# serializer version: 1
# name: test_yield_readings[lines0-\\s+-datetime_fieldnames0-datetime_formats0]
  list([
    dict({
      'reading': '20.54',
      'sensor': 'M(m/s)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 0),
    }),
    dict({
      'reading': '211.0',
      'sensor': 'D(deg)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 0),
    }),
    dict({
      'reading': '1.22',
      'sensor': 'SD(m/s)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 0),
    }),
    dict({
      'reading': '0.3',
      'sensor': 'DSD(deg)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 0),
    }),
    dict({
      'reading': '21.00',
      'sensor': 'Gust3s(m/s)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 0),
    }),
    dict({
      'reading': '11.9',
      'sensor': 'T(C)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 0),
    }),
    dict({
      'reading': '992.8',
      'sensor': 'PRE(hPa)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 0),
    }),
    dict({
      'reading': '0.15',
      'sensor': 'RiNumber',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 0),
    }),
    dict({
      'reading': '0.18',
      'sensor': 'VertM(m/s)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 0),
    }),
    dict({
      'reading': '21.02',
      'sensor': 'M(m/s)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 10),
    }),
    dict({
      'reading': '212.2',
      'sensor': 'D(deg)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 10),
    }),
    dict({
      'reading': '2.55',
      'sensor': 'SD(m/s)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 10),
    }),
    dict({
      'reading': '0.6',
      'sensor': 'DSD(deg)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 10),
    }),
    dict({
      'reading': '21.35',
      'sensor': 'Gust3s(m/s)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 10),
    }),
    dict({
      'reading': '11.8',
      'sensor': 'T(C)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 10),
    }),
    dict({
      'reading': '992.7',
      'sensor': 'PRE(hPa)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 10),
    }),
    dict({
      'reading': '0.29',
      'sensor': 'RiNumber',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 10),
    }),
    dict({
      'reading': '-0.09',
      'sensor': 'VertM(m/s)',
      'timestamp': datetime.datetime(2015, 12, 22, 0, 10),
    }),
  ])
# ---
